Gems/Atom/RHI/DX12/Code/Source/RHI/BufferD3D12MemoryAllocator.cpp

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright (c) Contributors to the Open 3D Engine Project.
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+#include <RHI/BufferD3D12MemoryAllocator.h>
+#include <RHI/Conversions.h>
+#include <RHI/Device.h>
+
+#include <AzCore/Casting/lossy_cast.h>
+
+namespace AZ
+{
+    namespace DX12
+    {
+        namespace Platform
+        {
+            D3D12_RESOURCE_STATES GetRayTracingAccelerationStructureResourceState();
+        }
+
+        void BufferD3D12MemoryAllocator::Init(const Descriptor& descriptor)
+        {
+            m_descriptor = descriptor;
+        }
+
+        void BufferD3D12MemoryAllocator::Shutdown()
+        {
+        }
+
+        void BufferD3D12MemoryAllocator::GarbageCollect()
+        {
+        }
+
+        BufferMemoryView BufferD3D12MemoryAllocator::Allocate(size_t sizeInBytes, size_t overrideSubAllocAlignment [[maybe_unused]])
+        {
+            AZ_PROFILE_FUNCTION(RHI);
+
+            RHI::BufferDescriptor bufferDescriptor;
+            bufferDescriptor.m_byteCount = azlossy_caster(sizeInBytes);
+            bufferDescriptor.m_bindFlags = m_descriptor.m_bindFlags;
+
+            D3D12_RESOURCE_STATES initialResourceState = ConvertInitialResourceState(m_descriptor.m_heapMemoryLevel, m_descriptor.m_hostMemoryAccess);
+            if (RHI::CheckBitsAny(m_descriptor.m_bindFlags, RHI::BufferBindFlags::RayTracingAccelerationStructure))
+            {
+                initialResourceState = Platform::GetRayTracingAccelerationStructureResourceState();
+            }
+
+            const D3D12_HEAP_TYPE heapType = ConvertHeapType(m_descriptor.m_heapMemoryLevel, m_descriptor.m_hostMemoryAccess);
+
+            MemoryView memoryView = m_descriptor.m_device->CreateD3d12maBuffer(bufferDescriptor, initialResourceState, heapType);
+            if (memoryView.IsValid())
+            {
+                const size_t sizeAllocated = memoryView.GetSize();
+
+                // Add the resident usage now that everything succeeded.
+                RHI::HeapMemoryUsage& heapMemoryUsage = *m_descriptor.m_getHeapMemoryUsageFunction();
+                heapMemoryUsage.m_totalResidentInBytes += sizeAllocated;
+                heapMemoryUsage.m_usedResidentInBytes += sizeAllocated;
+                heapMemoryUsage.m_uniqueAllocationBytes += sizeAllocated;
```

Comment:
Hmm. This is a little incorrect. Ideally these variables should be different
**m_totalResidentInBytes** is the total amount of memory allocated <---This seems incorrect. This number should represent how big a heap dx12ma allocated under th ehood.
**m_usedResidentInBytes** is the total amount of memory used <-------This seems correct
**m_uniqueAllocationBytes** is the total amount of unique memory <--This seems incorrect too. This may just be 0 in this case as I am assuming dx12ma is making placed allocations as oppose to commited allocations. Maybe there is a way to find out what type of allocation wass made and update this variable based on that.

Reply:
The problem is that d3d12ma shares heap space among pools, rather than create a memory page for each pool. So the best we can say is how much it is currently using for this pool.